### DISCUSSION TITLE
Updating to ASP.NET Core 3.1

### DISCUSSION THREAD

**Discussion by volkanceylan** [serenity-is/Serenity#5737] [Announcements]
Starting with 3.12 Serenity supports ASP.NET Core 3.1, and this is an upgrade guide for those who use .NET Core version and wish to upgrade to .NET Core 3.1. 

> Please note that this document and warnings below has nothing to do with those who use .NET MVC version of Serenity (although you are strongly recommended to migrate to .NET Core version, as ASP.NET MVC has no future).

**Important Warnings:**

- .NET Core 3 SDK is only supported with Visual Studio 2019 due to Microsoft constraints, so if you have to use an older version, don't upgrade Serenity, latest version you can use is **3.9.14**, you should stay at that version if you would like to use Visual Studio 2017 and .NET Core 2.2. If you upgraded by mistake, revert back to 3.9.14.
- ASP.NET 3 only works with .NET Core 3, it no longer supports desktop full .NET framework.
- Serenity versions after 3.9.14 will only support .NET Core 3.1 (and ASP.NET MVC, nothing changed with that)
- DotNetCliToolReference is deprecated so we have to use new dotnet tool system for sergen (as a local tool).

## Steps to Upgrade to ASP.NET Core 3.1:

- Start by ensuring your existing project is building fine, and there are no errors.
- Take a backup of entire project just in case (if not using version control)
- Make sure you have latest Visual Studio 2019 updates and .NET Core 3.1 SDK installed.
- Edit YourApp.Web.csproj in Visual Studio, and remove the ItemGroup shown below:
```xml
  <ItemGroup>
    <DotNetCliToolReference Include="Serenity.CodeGenerator" Version="3.9.14" />
  </ItemGroup>
```
- Replace the target framework:

```xml
<TargetFramework>netcoreapp2.2</TargetFramework>
```

with:
```xml
<TargetFramework>netcoreapp3.1</TargetFramework>
```

- remove the DotNetSergen line at top that looks like this (the first one with net461, not netcoreapp2.2):

```xml
    <DotNetSergen Condition=" '$(TargetFramework)' == 'net461' AND Exists('..\..\Serenity\Serenity.CodeGenerator\bin\$(Configuration)\$(TargetFramework)\dotnet-sergen.exe')">..\..\Serenity\Serenity.CodeGenerator\bin\$(Configuration)\$(TargetFramework)\dotnet-sergen.exe</DotNetSergen>
```

- Replace the second DotNetSergen line:

```xml
<DotNetSergen Condition=" '$(TargetFramework)' == 'netcoreapp2.2' AND Exists('..\..\Serenity\Serenity.CodeGenerator\bin\$(Configuration)\$(TargetFramework)\dotnet-sergen.dll')">dotnet ..\..\Serenity\Serenity.CodeGenerator\bin\$(Configuration)\$(TargetFramework)\dotnet-sergen.dll</DotNetSergen>
```

with:

```xml
<DotNetSergen Condition="Exists('..\..\Serenity\Serenity.CodeGenerator\bin\$(Configuration)\sergen.exe')">..\..\Serenity\Serenity.CodeGenerator\bin\$(Configuration)\sergen.exe</DotNetSergen>
```

- Let's also update TypeScript to 3.7.4 (not mandatory), replace this (assuming you are using 3.0 yet):

```xml
<TypeScriptToolsVersion>3.0</TypeScriptToolsVersion>
```

with:

```xml
<TypeScriptToolsVersion>3.7</TypeScriptToolsVersion>
```

And replace this line:

```xml
    <PackageReference Include="Microsoft.TypeScript.MSBuild" Version="3.1.1" />
```

with:

```xml
    <PackageReference Include="Microsoft.TypeScript.MSBuild" Version="3.7.4" />
```

- Remove the ItemGroup section that contains **TargetFramework == net461** (as desktop framework is no longer possible to use with ASP.NET Core 3:

```xml
  <ItemGroup Condition=" '$(TargetFramework)' == 'net461' ">
    <PackageReference Include="EPPlus" Version="4.5.2.1" />
   ...
    ... remove this ItemGroup and just this one with '$(TargetFramework)' == 'net461' 
    ...
    <PackageReference Include="Serenity.Scripts" Version="3.9.14" />
  </ItemGroup>
```

- Similarly remove the PropertyGroup for net461:

```xml
  <PropertyGroup Condition=" '$(TargetFramework)' == 'net461' ">
    <PlatformTarget>x64</PlatformTarget>
    <DefineConstants>$(DefineConstants);ASPNETCORE</DefineConstants>
  </PropertyGroup>
```

- As we are going to have just one target framework, find the section like below (you might have additional packages there, that's ok):

```xml
  <ItemGroup>
    <PackageReference Include="MailKit" Version="2.0.6" />
    <PackageReference Include="Microsoft.TypeScript.MSBuild" Version="3.7.4" />
    <PackageReference Include="Serenity.FluentMigrator.Runner" Version="1.6.904" />
    <PackageReference Include="Microsoft.Data.Sqlite" Version="2.0.0" />
  </ItemGroup>
  <ItemGroup Condition=" '$(TargetFramework)' == 'netcoreapp2.2' ">
    <PackageReference Include="EPPlus" Version="4.5.2.1" />
    <PackageReference Include="FastMember" Version="1.4.1" />
    <PackageReference Include="Microsoft.AspNetCore.App" />
    <PackageReference Include="Serenity.Web" Version="3.9.14" />
    <PackageReference Include="Serenity.Web.Assets" Version="3.9.2" />
    <PackageReference Include="Serenity.Scripts" Version="3.9.14" />
  </ItemGroup>
```

and remove the first ItemGroup closing tag, and ItemGroup opening tag with Condition so that it looks like this (we are just merging item groups and removing the unused condition):

```xml
  <ItemGroup>
    <PackageReference Include="MailKit" Version="2.0.6" />
    <PackageReference Include="Microsoft.TypeScript.MSBuild" Version="3.7.4" />
    <PackageReference Include="Serenity.FluentMigrator.Runner" Version="1.6.904" />
    <PackageReference Include="Microsoft.Data.Sqlite" Version="2.0.0" />
    <PackageReference Include="EPPlus" Version="4.5.2.1" />
    <PackageReference Include="FastMember" Version="1.4.1" />
    <PackageReference Include="Microsoft.AspNetCore.App" />
    <PackageReference Include="Serenity.Web" Version="3.9.14" />
    <PackageReference Include="Serenity.Web.Assets" Version="3.9.2" />
    <PackageReference Include="Serenity.Scripts" Version="3.9.14" />
  </ItemGroup>
```

- Remove the Microsoft.AspNetCore.App line:

```xml
    <PackageReference Include="Microsoft.AspNetCore.App" />
```

- Add these two packages to that ItemGroup

```xml
    <PackageReference Include="Microsoft.AspNetCore.Mvc.NewtonsoftJson" Version="3.1.0" />
    <PackageReference Include="Microsoft.AspNetCore.Mvc.Razor.RuntimeCompilation" Version="3.1.0" />
```

- Update Serenity packages to latest versions (current versions shown below):

```xml
    <PackageReference Include="Serenity.Pro.Scripts" Version="3.9.10" />
    <PackageReference Include="Serenity.Scripts" Version="3.9.14" />
    <PackageReference Include="Serenity.Web" Version="3.9.14" Condition="!Exists('..\..\Serenity\Serenity.Core\Serenity.Core.csproj')" />
    <PackageReference Include="Serenity.Web.Assets" Version="3.9.2" />
```

> The line with Serenity.Pro.Scripts is only for StartSharp users.

- Update TypeScript to 3.7.4:

```xml
    <PackageReference Include="Microsoft.TypeScript.MSBuild" Version="3.7.4"/>
```

- Optionally update other packages like MailKit, EPPlus, FastMember, Mapster (if you have them) to latest versions:

```xml
  <ItemGroup>
    <PackageReference Include="EPPlus" Version="4.5.3.2" />
    <PackageReference Include="FastMember" Version="1.5.0" />
    <PackageReference Include="MailKit" Version="2.4.1" />
    <PackageReference Include="Mapster" Version="4.1.1" />
    <PackageReference Include="Microsoft.Data.Sqlite" Version="3.1.0" />
    <PackageReference Include="StackExchange.Exceptional.AspNetCore" Version="2.1.0" />
  </ItemGroup>
```

> The lines with StackExchange.Exceptional and Mapster are only for StartSharp users.

Eventually your ItemGroup should look similar to this (for Serene):

```xml
  <ItemGroup>
    <PackageReference Include="EPPlus" Version="4.5.3.2" />
    <PackageReference Include="FastMember" Version="1.5.0" />
    <PackageReference Include="MailKit" Version="2.4.1" />
    <PackageReference Include="Microsoft.AspNetCore.Mvc.NewtonsoftJson" Version="3.1.0" />
    <PackageReference Include="Microsoft.AspNetCore.Mvc.Razor.RuntimeCompilation" Version="3.1.0" />
    <PackageReference Include="Microsoft.Data.Sqlite" Version="3.1.0" />
    <PackageReference Include="Microsoft.TypeScript.MSBuild" Version="3.7.4"/>
    <PackageReference Include="Serenity.FluentMigrator.Runner" Version="1.6.904" />
    <PackageReference Include="Serenity.Scripts" Version="3.9.14" />
    <PackageReference Include="Serenity.Web" Version="3.9.14" Condition="!Exists('..\..\Serenity\Serenity.Core\Serenity.Core.csproj')" />
    <PackageReference Include="Serenity.Web.Assets" Version="3.9.2" />
  </ItemGroup>
```

> You might also have extra packages there if you did install some yourself, so don't remove them, the above one is just a sample.

- Find this PropertyGroup just below the one above, and remove the condition and remove the AspNetCoreHostingModel line:

```xml
  <PropertyGroup Condition=" '$(TargetFramework)' == 'netcoreapp2.2' ">
    <DefineConstants>$(DefineConstants);COREFX;ASPNETCORE</DefineConstants>
    <AspNetCoreHostingModel>InProcess</AspNetCoreHostingModel>
  </PropertyGroup>
```

so that it looks like this:

```xml
  <PropertyGroup>
    <DefineConstants>$(DefineConstants);COREFX;ASPNETCORE</DefineConstants>
  </PropertyGroup>
```

- Replace the TransformMvcClientTypes section:

```xml
  <Target Name="TransformMvcClientTypes" BeforeTargets="BeforeBuild">
    <Exec Command="$(DotNetSergen) mvct" ContinueOnError="true" />
  </Target>
```

with this (insert dotnet tool restore line):

```xml
  <Target Name="TransformMvcClientTypes" BeforeTargets="BeforeBuild">
    <Exec Command="dotnet tool restore" ContinueOnError="true" />
    <Exec Command="$(DotNetSergen) mvct" ContinueOnError="true" />
  </Target>
```

- Open a command prompt, CD to project directory (with .CSPROJ file) and run the commands shown in order

```cmd
dotnet new tool-manifest
```

```cmd
dotnet tool install sergen
```

```cmd
dotnet restore
```

- Edit your Program.cs, make sure Program class looks like this:

```cs

using System.IO;
using Microsoft.AspNetCore.Hosting;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.Hosting;

namespace ..MyProject..
{
   public class Program
   {
        public static void Main(string[] args)
        {
            CreateHostBuilder(args).Build().Run();
        }

        public static IHostBuilder CreateHostBuilder(string[] args)
        {
            return Host.CreateDefaultBuilder(args)
                .ConfigureWebHostDefaults(webBuilder =>
                {
                    webBuilder.UseStartup<Startup>();
                })
                .ConfigureAppConfiguration((builderContext, config) =>
                {
                    config.AddJsonFile($"appsettings.machine.json", optional: true);
                });
        }
   }
}
```

- Edit your Initialization/Startup.cs, add these two using clauses:

```cs
using Microsoft.AspNetCore.Server.Kestrel.Core;
using Microsoft.Extensions.Hosting;
```

- Change Startup constructor from this:

```cs
        public Startup(IConfiguration configuration, IWebHostEnvironment hostEnvironment)
        {
            Configuration = configuration;
        }
```

to this:

```cs
        public Startup(IConfiguration configuration, IWebHostEnvironment hostEnvironment)
        {
            Configuration = configuration;
            HostEnvironment = hostEnvironment;
        }
```

- Add this HostEnvironment line under IConfiguration:

```cs
        public IConfiguration Configuration { get; }
        public IWebHostEnvironment HostEnvironment { get; }
```

- Before the services.AddMvc line in ConfigureServices method, add these lines:

```cs
            services.Configure<KestrelServerOptions>(options =>
            {
                options.AllowSynchronousIO = true;
            });

            services.Configure<IISServerOptions>(options =>
            {
                options.AllowSynchronousIO = true;
            });
```

- Replace the services.AddMvc part:

```cs
            services.AddMvc(options =>
            {
                options.Filters.Add(new AutoValidateAntiforgeryTokenAttribute());
                options.Filters.Add(typeof(AntiforgeryCookieResultFilterAttribute));
                options.ModelBinderProviders.Insert(0, new ServiceEndpointModelBinderProvider());
                options.Conventions.Add(new ServiceEndpointActionModelConvention());
            })
            .AddJsonOptions(options =>
            {
                options.SerializerSettings.ContractResolver = new DefaultContractResolver();
            });

```
with:

```cs
            var builder = services.AddControllersWithViews(options =>
            {
                options.Filters.Add(typeof(AutoValidateAntiforgeryTokenAttribute));
                options.Filters.Add(typeof(AntiforgeryCookieResultFilterAttribute));
                options.ModelBinderProviders.Insert(0, new ServiceEndpointModelBinderProvider());
                options.Conventions.Add(new ServiceEndpointActionModelConvention());
            }).AddNewtonsoftJson(options =>
            {
                options.SerializerSettings.ContractResolver = new DefaultContractResolver();
            });
            
            if (HostEnvironment.IsDevelopment())
                builder.AddRazorRuntimeCompilation();
```

- Replace IHostingEnvironment with IWebHostEnvironment in Configure method:

```cs
public void Configure(IApplicationBuilder app, IWebHostEnvironment env, ILoggerFactory loggerFactory, IAntiforgery antiforgery)
```
- After app.UseStaticFiles() line add app.UseRouting(), and after app.UseAuthentication add app.UseAuthorization:

```cs
            app.UseStaticFiles();
            app.UseRouting(); // insert this
            app.UseAuthentication();
            app.UseAuthorization(); // insert this
```

- Replace app.UseMvc line:

```cs
	    app.UseMvc(routes =>
            {
            });
```

with:

```
            app.UseEndpoints(endpoints => {
                endpoints.MapControllers();
            });
```

- Edit Initialization/DataMigrations.cs, SergenEdpoint.cs, and EmailHelper.cs. Replace **IHostingEnvironment** with **IWebHostEnvironment** (there should be five or so).

- Rebuild the solution and make sure there are no errors.

- It was a long upgrade guide but unfortunately there are too many changes in ASP.NET Core and .NET Core themselves. You may look at following guides to see how complex it is (particulary the first one):

https://docs.microsoft.com/en-us/aspnet/core/migration/22-to-30?view=aspnetcore-3.1&tabs=visual-studio

https://docs.microsoft.com/en-us/aspnet/core/migration/30-to-31?view=aspnetcore-3.1&tabs=visual-studio

**Comment by marcobisio** [comment]
Great job @volkanceylan !!

**Comment by JohnRanger** [comment]
(Now we know what you did over Chrismas... :-))

Good to see you back in full action :-)

Wish you a great, successful 2020.

With kind regards,

John

**Comment by kilroyFR** [comment]
Great news, 
another good news is my renewal subscription for another year with this awesome Framework

**Comment by haoas** [comment]
great！has adminLTE also upgraded to 3？

**Comment by Estrusco** [comment]
Great job! Thank you very much @volkanceylan !

**Comment by edwardch** [comment]
Thank you @volkanceylan for your hard work and progress on Serenity - my favourite framework!

I followed this guide on my project and I'm getting a warning and error about Razor. Maybe this is just something that I've done?

```
1>------ Build started: Project: MyApp.Web, Configuration: Debug Any CPU ------
1>Tool 'sergen' (version '3.12.0') was restored. Available commands: sergen
1>
1>Restore was successful.
1>Transforming MVC at: C:\source\repos\MyApp\MyApp\MyApp.Web\Imports\MVC
1>Transforming ClientTypes at: C:\source\repos\MyApp\MyApp\MyApp.Web\Imports\ClientTypes
1>C:\Program Files\dotnet\sdk\3.1.100\Sdks\Microsoft.NET.Sdk.Razor\build\netstandard2.0\Sdk.Razor.CurrentVersion.targets(404,5): warning RAZORSDK1006: Detected Razor language version downgrade. This is typically caused by a reference to the Microsoft.AspNetCore.Razor.Design package. Consider removing this package reference.
1>C:\Program Files\dotnet\sdk\NuGetFallbackFolder\microsoft.aspnetcore.razor.design\2.1.1\build\netstandard2.0\Microsoft.AspNetCore.Razor.Design.CodeGeneration.targets(69,5): error : The project C:\source\repos\MyApp\MyApp\MyApp.Web must provide a value for Configuration.
1>Done building project "MyApp.Web.csproj" -- FAILED.
========== Build: 0 succeeded, 1 failed, 0 up-to-date, 0 skipped ==========
```

It mentions removing the `Microsoft.AspNetCore.Razor.Design` package, but I can't seem to find this package or reference to it.

My environment before upgrade:
- Win10, VS2019
- .NET Core 2.2
- Serenity 3.9.14
- StartSharp 3.9.14 (folder says 3.9.10 though?)
- App was building and running fine

Has anyone else migrated and/or run into this issue?

Thank you,
Edward

**Comment by mustafalotiya** [comment]
Hi @volkanceylan 

Thank you for update.

Kindly request you to please prepare and share Migration tutorial of  **Serenity MVC to .NET core version.**

Thanks.

**Comment by volkanceylan** [comment]
@edwardch need your CSPROJ file references section

**Comment by edwardch** [comment]
Here we go
```xml
  <ItemGroup>
    <PackageReference Include="Azure.Storage.Blobs" Version="12.1.0" />
    <PackageReference Include="CsvHelper" Version="12.2.3" />
    <PackageReference Include="DocumentFormat.OpenXml" Version="2.9.1" />
    <PackageReference Include="Hangfire.AspNetCore" Version="1.7.8" />
    <PackageReference Include="HangFire.Core" Version="1.7.8" />
    <PackageReference Include="HangFire.SqlServer" Version="1.7.8" />
    <PackageReference Include="MailKit" Version="2.4.1" />
    <PackageReference Include="Mapster" Version="4.1.1" />
    <PackageReference Include="Microsoft.TypeScript.MSBuild" Version="3.7.4" />
    <PackageReference Include="Microsoft.VisualStudio.Azure.Containers.Tools.Targets" Version="1.9.10" />
    <PackageReference Include="Serenity.FluentMigrator.Runner" Version="1.6.904" />
    <PackageReference Include="Serenity.Pro.Scripts" Version="3.9.10" />
    <PackageReference Include="Spire.XLS" Version="9.12.20" />
    <PackageReference Include="StackExchange.Exceptional.AspNetCore" Version="2.1.0" />
    <PackageReference Include="Microsoft.Data.Sqlite" Version="3.1.0" />
    <PackageReference Include="Telerik.Reporting" Version="13.0.19.222" />
    <PackageReference Include="Telerik.Reporting.OpenXmlRendering" Version="13.0.19.222" />
    <PackageReference Include="Telerik.Reporting.Services.AspNetCore" Version="13.0.19.222" />
    <PackageReference Include="EPPlus" Version="4.5.3.2" />
    <PackageReference Include="FastMember" Version="1.5.0" />
    <PackageReference Include="Serenity.Web" Version="3.12.0" Condition="!Exists('..\..\Serenity\Serenity.Core\Serenity.Core.csproj')" />
    <PackageReference Include="Serenity.Web.Assets" Version="3.9.2" />
    <PackageReference Include="Serenity.Scripts" Version="3.12.0" />
    <PackageReference Include="Microsoft.AspNetCore.Mvc.NewtonsoftJson" Version="3.1.0" />
    <PackageReference Include="Microsoft.AspNetCore.Mvc.Razor.RuntimeCompilation" Version="3.1.0" />	
  </ItemGroup>
```

**Comment by mustafalotiya** [comment]
> Here we go
> 
> ```
>   <ItemGroup>
>     <PackageReference Include="Azure.Storage.Blobs" Version="12.1.0" />
>     <PackageReference Include="CsvHelper" Version="12.2.3" />
>     <PackageReference Include="DocumentFormat.OpenXml" Version="2.9.1" />
>     <PackageReference Include="Hangfire.AspNetCore" Version="1.7.8" />
>     <PackageReference Include="HangFire.Core" Version="1.7.8" />
>     <PackageReference Include="HangFire.SqlServer" Version="1.7.8" />
>     <PackageReference Include="MailKit" Version="2.4.1" />
>     <PackageReference Include="Mapster" Version="4.1.1" />
>     <PackageReference Include="Microsoft.TypeScript.MSBuild" Version="3.7.4" />
>     <PackageReference Include="Microsoft.VisualStudio.Azure.Containers.Tools.Targets" Version="1.9.10" />
>     <PackageReference Include="Serenity.FluentMigrator.Runner" Version="1.6.904" />
>     <PackageReference Include="Serenity.Pro.Scripts" Version="3.9.10" />
>     <PackageReference Include="Spire.XLS" Version="9.12.20" />
>     <PackageReference Include="StackExchange.Exceptional.AspNetCore" Version="2.1.0" />
>     <PackageReference Include="Microsoft.Data.Sqlite" Version="3.1.0" />
>     <PackageReference Include="Telerik.Reporting" Version="13.0.19.222" />
>     <PackageReference Include="Telerik.Reporting.OpenXmlRendering" Version="13.0.19.222" />
>     <PackageReference Include="Telerik.Reporting.Services.AspNetCore" Version="13.0.19.222" />
>     <PackageReference Include="EPPlus" Version="4.5.3.2" />
>     <PackageReference Include="FastMember" Version="1.5.0" />
>     <PackageReference Include="Serenity.Web" Version="3.12.0" Condition="!Exists('..\..\Serenity\Serenity.Core\Serenity.Core.csproj')" />
>     <PackageReference Include="Serenity.Web.Assets" Version="3.9.2" />
>     <PackageReference Include="Serenity.Scripts" Version="3.12.0" />
>     <PackageReference Include="Microsoft.AspNetCore.Mvc.NewtonsoftJson" Version="3.1.0" />
>     <PackageReference Include="Microsoft.AspNetCore.Mvc.Razor.RuntimeCompilation" Version="3.1.0" />	
>   </ItemGroup>
> ```

Hi @edwardch ,

found that you are using telerik reporting.. Can you please help me on implementation of telerik reporting on serenity project.

skype Id: mustafalotiya@gmail.com

Regards,
Mustufa

**Comment by edwardch** [comment]
@mustafalotiya I have this working in .NET MVC, and working 95% in .NET Core. 

The catch in .NET Core is that the reports are not finding my named connection string. Reports that do not use a connection string do work though. I got paid-for support from Volkan to help me set this up initially for the .NET MVC version, and I plan on fixing that connection string issue first before creating a tutorial on how to do that. 

I will post an issue to announce when the tutorial is complete.

**Comment by mustafalotiya** [comment]
> @mustafalotiya I have this working in .NET MVC, and working 95% in .NET Core.
> 
> The catch in .NET Core is that the reports are not finding my named connection string. Reports that do not use a connection string do work though. I got paid-for support from Volkan to help me set this up initially for the .NET MVC version, and I plan on fixing that connection string issue first before creating a tutorial on how to do that.
> 
> I will post an issue to announce when the tutorial is complete.

As of now my all developments in Serenity MVC framework and looking for good reporting tool. it would be great help.

**Comment by RouteTech** [comment]
@edwardch I had a similar problem but it only showed up after I tried to publish the app, so in my case it was related to my publish setup.

**Comment by edwardch** [comment]
@RouteTech I haven't got any publish profiles, but I'm interested to know what you had different in your publish profile to what was on your local/dev environment?

**Comment by volkanceylan** [comment]
@edwardch Telerik.Reporting.Services.AspNetCore 13.0.19.222 probably has reference to ASP.NET Core 2.0, as the package is private i can't check deps

**Comment by volkanceylan** [comment]
Telerik.Reporting for example is listed in NuGet only for 13.2.19.1030 and it is dependent on .NETCoreApp2.0 so i guess your version is also. You need a .NETCoreApp3.1 version

**Comment by RouteTech** [comment]
@edwardch I had it in two situations, the first was one of the parameters in the existing profile was targetting the 2.2 framework and I didn't notice. The second was when I was trying to "Make a tiny .NET Core 3.0 entirely self-contained single executable"

**Comment by edwardch** [comment]
> @edwardch Telerik.Reporting.Services.AspNetCore 13.0.19.222 probably has reference to ASP.NET Core 2.0, as the package is private i can't check deps

Thank you @volkanceylan, that was exactly the issue! Project now builds successfully.

It won't run yet though because when adding the Hangfire service in `Startup.cs` it uses `SqlConnections.GetConnectionString("Default").ConnectionString`, and I get an error `System.InvalidOperationException: 'No connection string with key Default in configuration file!'`

I will have another look tomorrow as I need to go out now.

**Comment by edwardch** [comment]
After the upgrade I could not get the connection string for Hangfire to work in my Startup.cs (mentioned above).

```cs
SqlConnections.GetConnectionString("Hangfire")
``` 
This returns `System.InvalidOperationException: 'No connection string with key Hangfire in configuration file!'`

To get around the error I changed it to this:
```cs
Configuration.GetValue<string>("Data:Hangfire:ConnectionString")
```

@volkanceylan This worked, and I may be wrong, but this seems like a non-standard way of defining the connection string, I'd be interested to know why it is done this way?

My other option was to define a ConnectionStrings section in my `appsettings.json` file like I've previously seen online:
```json
  "Data": {
    "Default": {
      "ConnectionString": "Server=(localdb)\\MsSqlLocalDB;Database=Serene_Default_v1;Integrated Security=true",
      "ProviderName": "System.Data.SqlClient"
    }
  },
  "ConnectionStrings": {
    "Hangfire": "Server=(localdb)\\MsSqlLocalDB;Database=Serene_Default_v1;Integrated Security=true",
    "ProviderName": "System.Data.SqlClient"
  },
  "Logging": "..."
```

Then refer to this connection string in `Startup.cs` like so:
```cs
Configuration.GetConnectionString("Hangfire")
```
I have it working, though I see `SqlConnections.GetConnectionString(databaseKey)` in `DataMigration.cs` which works, so any feedback would be useful.

Thank you in advance.

**Comment by volkanceylan** [comment]
There was standard no connection string section in appsettings for .NET Core when we migrated, there wasn't even a provider system, so no provider name, so we designed our own connection string system.

There wasn't a change to this, i don't know why you are getting error. Anyway you may use Configuration.GetValue that's fine.

**Comment by edwardch** [comment]
Ah that makes sense, thank you for that feedback. I'm happy with the solution and looking forward to getting my upgrades into production!

**Comment by she-ll-b-rite** [comment]
Thanks for the upgrade guide. Worked great, except the build task "CompileTSC" keeps giving the following error message when building:
```
1>The directory name is invalid.
1>C:\Users\xxx\Documents\Dev\xxx\src\xxx.Web\xxx.Web.csproj(88,5): warning MSB3073: The command ""C:\Users\xxx\.nuget\packages\microsoft.typescript.msbuild\3.7.4\build\..\tools\tsc\tsc.js\tsc" -p ./tsconfig.json" exited with code 267.
```

The build task in project file looks like this:
```
  <Target Name="CompileTSC" AfterTargets="AfterBuild" Inputs="@(CompileTSCInputs)" Outputs="wwwroot\Scripts\site\xxx.Web.js">
    <Message Text="Executing TypeScript compiler..." Importance="high"></Message>
    <Exec Command="&quot;$(TscToolPath.Replace('build\\..\tools\', 'tools\'))\tsc&quot; -p ./tsconfig.json" ContinueOnError="true" />
  </Target>
```

on investigation looks like "tsc.exe" is not part of 3.7.4 Typescript SDK anymore.

Any suggestions for fixing?

**Comment by volkanceylan** [comment]
@she-ll-b-rite you should first follow other upgrade guides:

https://serenity.is/docs/startsharp/updating-startsharp

**Comment by vitorsilva** [comment]
hi, i'm trying serenity for the first time. 
downloaded it and installed it, upgraded it to 3.1, everything builds with no errors, but when i start the application the site opens but it doesn't show the login screen

![image](https://user-images.githubusercontent.com/35258/73793964-58002c80-479f-11ea-94e2-45cb2bf63312.png)

it's probably a newbie mistake... but any help would be great

**Comment by she-ll-b-rite** [comment]
> @she-ll-b-rite you should first follow other upgrade guides:
> 
> https://serenity.is/docs/startsharp/updating-startsharp

My bad. Thanks 👍

**Comment by TheoKand** [comment]
> > @she-ll-b-rite you should first follow other upgrade guides:
> > https://serenity.is/docs/startsharp/updating-startsharp
> 
> Hi @volkanceylan ,
> 
> I have the same issue with @she-ll-b-rite . I get this error
> 
> _An error occurred while starting the application. JsonReaderException: The input does not contain any JSON tokens. Expected the input to start with a valid JSON token, when isFinalBlock is true. LineNumber: 0 | BytePositionInLine: 0._
> 
> inside Program.cs > CreateHostBuilder(args).Build().Run();
> 
> and I am not using StartSharp. Any ideas?
> 
> Thanks in advance,
> Theo

To anyone that has the same problem, I resolved this by adding an empty object {} inside secrets.json

**Comment by lechu90ns** [comment]
Does anyone know if dotnet sergen servertypings should be still inside building process? I guess I am missing it somehow.

Thanks!

**Comment by lechu90ns** [comment]
> Does anyone know if dotnet sergen servertypings should be still inside building process? I guess I am missing it somehow.
> 
> Thanks!

Nvm, I replaced the wrong build tag inside web.csproj.

**Comment by ga5tan** [comment]
@vitorsilva Hi
were you able to fix it?
I am getting the same error. seems like no JS scripts are being loaded from IIS Express

**Comment by ga5tan** [comment]
so how do I run Sergen now ?

**Comment by wnchong** [comment]
> 
> 
> hi, i'm trying serenity for the first time.
> downloaded it and installed it, upgraded it to 3.1, everything builds with no errors, but when i start the application the site opens but it doesn't show the login screen
> 
> ![image](https://user-images.githubusercontent.com/35258/73793964-58002c80-479f-11ea-94e2-45cb2bf63312.png)
> 
> it's probably a newbie mistake... but any help would be great

I have the same problem as @vitorsilva. VS2019 - Serene - ASP.NET Core 2.2 -> 3.1.

**Comment by ga5tan** [comment]
@zhiwenchong  for me, this was caused by the fact, that no all the files were installed.
Check browser console (F12) for errors. bunch of JS errorrs. Even Bootstrap is missing
Try to generate solution again (that helped me)

**Comment by lanhouzi** [comment]
Nice job!

**Comment by PassionateDeveloper86** [comment]
Hi,

I just updated from 2.2 to 3.1.
Everything went 99% fluid, very nice!

Sadly I hade a problem with the IF COREFX and IF !COREFX Parts.
I had to exectly reversed as they where before to be able to startup and run my application without error.

I don't know why and what this is doing, but now (for now) its workling.

Any ideas?

**Comment by OpenlabWA** [comment]
Thank you sooo much for taking the time to write this up, it was effortless to move across!
